probe-rs/src/probe/fake_probe.rs

```diff
@@ -80,90 +86,96 @@ impl SwdSequence for &mut MockCore {
 }
 
 impl MemoryInterface<ArmError> for &mut MockCore {
-    fn read_8(&mut self, _address: u64, _data: &mut [u8]) -> Result<(), ArmError> {
-        todo!()
+    fn read_8(&mut self, address: u64, data: &mut [u8]) -> Result<(), ArmError> {
+        self.arm_read_handler.as_ref().unwrap()(address, data)
     }
 
-    fn read_16(&mut self, _address: u64, _data: &mut [u16]) -> Result<(), ArmError> {
-        todo!()
+    fn read_16(&mut self, address: u64, data: &mut [u16]) -> Result<(), ArmError> {
+        self.arm_read_handler.as_ref().unwrap()(address, unsafe {
+            std::slice::from_raw_parts_mut(data.as_mut_ptr() as *mut u8, data.len() * 2)
+        })
     }
 
     fn read_32(&mut self, address: u64, data: &mut [u32]) -> Result<(), ArmError> {
-        for (i, val) in data.iter_mut().enumerate() {
-            let address = address + (i as u64 * 4);
-
-            match address {
-                // DHCSR
-                Dhcsr::ADDRESS_OFFSET => {
-                    let mut dhcsr: u32 = self.dhcsr.into();
+        match address {
```

Comment:
Why did you change this to do a single read?

Reply:
The reason was that I wanted the callback handle the whole slice `&mut [u8]` rather than have it handle only a single value `&mut u8`. Looking at the implementation with fresh eyes, there are two edge cases where the DHCSR register isn't properly set:
- The user request something in the `0xE0000000..=0xE00FFFFF` range, and the final slice would contain the DHCSR value, but the address doesn't start at the DHCSR register.
- The use requests something outside the previously mentioned range, but is big enough to include the DHCSR register.

For the first edge case, it would make sense for the DHCSR register to be set properly, but for the second one it's a bit more tricky. We could determine if the user-requested slice intersects that range, and have the callback receive the modified slice so it's doesn't cover that area (note: this could mean 2 calls to the callbacks if the range is in the middle of the user's slice), or, simply let the user callback handle that range itself.

My gut reaction would be to patch the first edge case, but let the second one be "undefined behaviour" (documented, of course). What do you think?